[Delegation Toolkit] Deploy smart account guide (#2270)
MetaMask/metamask-docs: delegation-toolkit/guides/smart-accounts/deploy-smart-account.md

---
description: Learn how to deploy a MetaMask smart account.
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Deploy a smart account

> You can deploy MetaMask Smart Accounts in two different ways. You can either deploy the smart account automatically when sending 
> the first user operation, or use a manual approach.

Reviewer: ```suggestion
You can deploy MetaMask Smart Accounts in two different ways. You can either deploy a smart account automatically when sending 
the first user operation, or manually deploy the account.
```

## Prerequisites

- [Install and set up the Delegation Toolkit.](../../get-started/install.md)
- [Configure the Delegation Toolkit.](../configure.md)

> - [Create a MetaMask smart account](create-smart-account.md) 

Reviewer: ```suggestion
- [Create a MetaMask smart account.](create-smart-account.md) 
```

> ## Sending first user operation
> 
> Whenever you send the first user operation, it checks whether the smart account is already deployed. If the account 
> is not deployed, the `initCode` is added to the user operation to ensure the smart account is deployed within the 
> same operation. Internally, the `initCode` is encoded using the factory and factory data. 

Reviewer: ```suggestion
## Deploy with the first user operation

When you send the first user operation from a smart account, the Delegation Toolkit checks whether the account is already deployed. If the account 
is not deployed, the toolkit adds the `initCode` to the user operation to deploy the account within the 
same operation. Internally, the `initCode` is encoded using the `factory` and `factoryData`. 
```

const userOperationHash = await bundlerClient.sendUserOperation({
  account: smartAccount,
  calls: [
    {
      to: "0x1234567890123456789012345678901234567890",
      value: parseEther("1")
    }
  ],
  maxFeePerGas,
  maxPriorityFeePerGas
});
```

</TabItem>

<TabItem value="config.ts">

```typescript
import { createPublicClient, http } from "viem";
import { createBundlerClient } from "viem/account-abstraction";
import { sepolia as chain } from "viem/chains";
import { generatePrivateKey, privateKeyToAccount } from "viem/accounts";
import { 
  Implementation, 
  toMetaMaskSmartAccount,
} from "@metamask/delegation-toolkit";

const publicClient = createPublicClient({
  chain,
  transport: http()
});

const privateKey = generatePrivateKey(); 
const account = privateKeyToAccount(privateKey);

export const smartAccount = await toMetaMaskSmartAccount({
  client: publicClient,
  implementation: Implementation.Hybrid,
  deployParams: [account.address, [], [], []],
  deploySalt: "0x",
  signatory: { account },
});

export const bundlerClient = createBundlerClient({
  client: publicClient,
  transport: http("https://public.pimlico.io/v2/11155111/rpc")
});
```

</TabItem>
</Tabs>

> ## Manual approach
> 
> To use the manual approach, you can call the MetaMask Smart Account [`getFactoryArgs`](../../reference/api/smart-account#getfactoryargs)
> method to retrieve the `factory` and `factoryData`. This allows you to use a relay account to sponsor the deployment without needing a paymaster. 

Reviewer: ```suggestion
## Deploy manually

To deploy a smart account manually, call the [`getFactoryArgs`](../../reference/api/smart-account.md#getfactoryargs)
method from the smart account to retrieve the `factory` and `factoryData`. This allows you to use a relay account to sponsor the deployment without needing a paymaster. 
```

<Tabs>
<TabItem value="example.ts">

```typescript
import { walletClient, smartAccount } from "./config.ts";

const { factory, factoryData } = await smartAccount.getFactoryArgs();

// Deploy smart account using relay account.
const hash = await walletClient.sendTransaction({
  to: factory,
  data: factoryData,
})
```

</TabItem>

<TabItem value="config.ts">

```typescript
import { createPublicClient, createWalletClient, http } from "viem";
import { sepolia as chain } from "viem/chains";
import { generatePrivateKey, privateKeyToAccount } from "viem/accounts";
import { 
  Implementation, 
  toMetaMaskSmartAccount,
} from "@metamask/delegation-toolkit";

const publicClient = createPublicClient({
  chain,
  transport: http()
});

const privateKey = generatePrivateKey(); 
const account = privateKeyToAccount(privateKey);

export const smartAccount = await toMetaMaskSmartAccount({
  client: publicClient,
  implementation: Implementation.Hybrid,
  deployParams: [account.address, [], [], []],
  deploySalt: "0x",
  signatory: { account },
});

const relayAccountPrivateKey = "0x121..";

> const relayAccount = privateKeyToAccount(relayAccountPrivateKey)

Author: Ignoring this

export const walletClient = createWalletClient({
  account: relayAccount,
  chain,
  transport: http()
})
```

</TabItem>
</Tabs>

## Next steps

> - See [Send a user operation](send-user-operation.md) to learn how to send user operations. 
> - See [Send a gasless transaction](send-gasless-transaction.md) to learn how to use gasless paymaster.

Reviewer: ```suggestion
- Learn more about [sending user operations](send-user-operation.md).
- To sponsor gas for end users, see how to [send a gasless transaction](send-gasless-transaction.md).
```
